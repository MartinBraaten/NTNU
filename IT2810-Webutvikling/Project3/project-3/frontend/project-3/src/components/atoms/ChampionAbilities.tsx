//Display a champions abilities and passive in tabs with its description
import * as React from "react";
import Tabs from "@mui/material/Tabs";
import Tab from "@mui/material/Tab";
import { Box, Container, ThemeProvider, Typography } from "@mui/material";
import { fieldTheme } from "../themes/themes";
import { Spell } from "../../App";

//Inspired by https://mui.com/material-ui/react-tabs/

interface IChampionAbilities {
    passive: string[]; //eks: ["name of ability", "ability description", "ability image"]
    q: Spell;
    w: Spell;
    e: Spell;
    r: Spell;
}

interface TabPanelProps {
    children?: React.ReactNode;
    index: number;
    value: number;
}

//Panels that are shown for each tab
function TabPanel(props: TabPanelProps) {
    const { children, value, index, ...other } = props;

    return (
        <ThemeProvider theme={fieldTheme}>
            <div
                role="tabpanel"
                hidden={value !== index}
                id={`simple-tabpanel-${index}`}
                aria-labelledby={`simple-tab-${index}`}
                {...other}
            >
                {value === index && (
                    <Box sx={{ p: 3, maxWidth: "50vw" }}>
                        <Typography>{children}</Typography>
                    </Box>
                )}
            </div>
        </ThemeProvider>
    );
}

//Component that displays a champions 4 abilities and passive in tabs
export default function ChampionAbilities({
    passive,
    q,
    w,
    e,
    r,
}: IChampionAbilities) {
    const [value, setValue] = React.useState(0);

    const handleChange = (event: React.SyntheticEvent, newValue: number) => {
        setValue(newValue);
    };

    return (
        <Container sx={{ margin: "20px 40px" }}>
            <ThemeProvider theme={fieldTheme}>
                <h1 style={{ color: "white", fontSize: "60px" }}>Abilities</h1>
                <Tabs
                    value={value}
                    onChange={handleChange}
                    aria-label="icon position tabs example"
                    selectionFollowsFocus
                    id="abilityTabs"
                >
                    <Tab
                        sx={{ color: "white" }}
                        id="passiveTab"
                        icon={
                            <img
                                src={
                                    "http://ddragon.leagueoflegends.com/cdn/12.20.1/img/passive/" +
                                    passive[2]
                                }
                                alt="passive"
                            />
                        }
                        label="Passive"
                    ></Tab>
                    <Tab
                        sx={{ color: "white" }}
                        id="qTab"
                        icon={
                            <img
                                src={
                                    "http://ddragon.leagueoflegends.com/cdn/12.20.1/img/spell/" +
                                    q.image
                                }
                                alt="q"
                            />
                        }
                        label="Q"
                    />
                    <Tab
                        sx={{ color: "white" }}
                        id="wTab"
                        icon={
                            <img
                                src={
                                    "http://ddragon.leagueoflegends.com/cdn/12.20.1/img/spell/" +
                                    w.image
                                }
                                alt="w"
                            />
                        }
                        label="W"
                    />
                    <Tab
                        sx={{ color: "white" }}
                        id="eTab"
                        icon={
                            <img
                                src={
                                    "http://ddragon.leagueoflegends.com/cdn/12.20.1/img/spell/" +
                                    e.image
                                }
                                alt="e"
                            />
                        }
                        label="E"
                    />
                    <Tab
                        sx={{ color: "white" }}
                        id="rTab"
                        icon={
                            <img
                                src={
                                    "http://ddragon.leagueoflegends.com/cdn/12.20.1/img/spell/" +
                                    r.image
                                }
                                alt="r"
                            />
                        }
                        label="R"
                    />
                </Tabs>
            </ThemeProvider>
            <TabPanel value={value} index={0}>
                <h2 style={{ color: "#cfb53b" }}>{passive[0]}</h2>
                <p>{passive[1]}</p>
            </TabPanel>
            <TabPanel value={value} index={1}>
                <h2 style={{ color: "#cfb53b" }}>{q.name}</h2>
                <p>{q.description}</p>
            </TabPanel>
            <TabPanel value={value} index={2}>
                <h2 style={{ color: "#cfb53b" }}>{w.name}</h2>
                <p>{w.description}</p>
            </TabPanel>
            <TabPanel value={value} index={3}>
                <h2 style={{ color: "#cfb53b" }}>{e.name}</h2>
                <p>{e.description}</p>
            </TabPanel>
            <TabPanel value={value} index={4}>
                <h2 style={{ color: "#cfb53b" }}>{r.name}</h2>
                <p>{r.description}</p>
            </TabPanel>
        </Container>
    );
}
