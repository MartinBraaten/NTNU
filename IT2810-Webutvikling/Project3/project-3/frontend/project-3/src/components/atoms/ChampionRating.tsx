//Component that shows champion rating (and lets you rate a champion on click)

import { Box, Rating, Typography } from "@mui/material";
import "@fontsource/roboto/300.css";
import { useEffect, useState } from "react";
import StarIcon from "@mui/icons-material/Star";
import { gql, useMutation } from "@apollo/client";

//inspired by https://mui.com/material-ui/react-rating/
interface IChampionRating {
    name: string;
    review: number[];
}

//The different ratings a champion can have in LoL terms
const tiers: { [index: string]: string } = {
    1: "D-tier",
    2: "C-tier",
    3: "B-tier",
    4: "A-tier",
    5: "S-tier",
};

//posts a new rating to the database
const CHAMPIONMUTATION = gql`
    mutation Mutation($name: String!, $rateChampionInput: RateChampionInput) {
        addRating(name: $name, RateChampionInput: $rateChampionInput)
    }
`;

//Component for rating a champion
export default function ChampionRating({ review, name }: IChampionRating) {
    //Calculate the average off all the champions ratings stored in the database
    function getAverageRating() {
        let sum = 0;
        for (let i = 0; i < review.length; i++) {
            sum += review[i];
        }
        const average = sum / review.length;
        const roundedAverage =
            Math.round((average + Number.EPSILON) * 100) / 100;
        return roundedAverage;
    }

    //Add the new rating to the average
    function getNewAverageRating(newValuw: number) {
        let sum = 0;
        for (let i = 0; i < review.length; i++) {
            sum += review[i];
        }
        sum += newValuw;
        const average = sum / (review.length + 1);
        const roundedAverage =
            Math.round((average + Number.EPSILON) * 100) / 100;
        return roundedAverage;
    }

    const [averageRating, setRating] = useState<number>(getAverageRating());
    const [value, setValue] = useState<number | null>(getAverageRating());
    const [hover, setHover] = useState(-1);
    const [disabled, setDisabled] = useState<boolean>(false);

    // https://www.apollographql.com/docs/react/data/mutations/#example
    const [giveRating, { data, loading, error }] = useMutation(CHAMPIONMUTATION);

    //Disable the component when a rating is done for a champion
    const session = sessionStorage.getItem(name);
    useEffect(() => {
        if (session != null) {
            setDisabled(true);
        }
    }, [session]);

    return (
        <Box
            sx={{
                display: "flex",
                flexDirection: "column",
                margin: "auto",
            }}
        >
            <Box
                sx={{
                    display: "flex",
                    alignItems: "center",
                    gap: "10px",
                    color: "#fff",
                }}
            >
                <Rating
                    role="ratingComponent"
                    sx={{
                        color: "#cfb53b",
                    }}
                    value={value}
                    precision={1}
                    disabled={disabled}
                    onChange={(event, newValue) => {
                        setValue(newValue);
                        setDisabled(true);
                        if (newValue != null) {
                            setRating(getNewAverageRating(newValue)); // update average rating
                        }
                        sessionStorage.setItem(name, "done");
                        // giveRating mutates the database, adding a rating to the champion
                        giveRating({
                            variables: {
                                name: name,
                                rateChampionInput: { review: newValue },
                            },
                        });
                    }}
                    onChangeActive={(event, newHover) => {
                        setHover(newHover);
                    }}
                    emptyIcon={
                        <StarIcon
                            style={{ opacity: 0.55 }}
                            fontSize="inherit"
                        />
                    }
                    defaultValue={getAverageRating()}
                />
                <Typography
                    role="ratingText"
                    color="#fff"
                    align="left"
                    variant="h6"
                >
                    ({averageRating})
                </Typography>
            </Box>
            {value !== null && (
                <Box
                    sx={{
                        width: "120px",
                        color: "#fff",
                        fontSize: "18px",
                        textAlign: "center",
                    }}
                >
                    {tiers[hover !== -1 ? hover : value]}
                </Box>
            )}
        </Box>
    );
}
