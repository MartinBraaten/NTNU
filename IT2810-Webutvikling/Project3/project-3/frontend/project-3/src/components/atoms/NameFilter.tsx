//Filter for filtering on champion name
import {
    IconButton,
    InputAdornment,
    TextField,
    ThemeProvider,
} from "@mui/material";
import Box from "@mui/material/Box";
import { fieldTheme } from "../themes/themes";
import SearchIcon from "@mui/icons-material/Search";

export interface IFilter {
    change: boolean | undefined;
    setChange: (change: boolean) => void;
}

//Component for searching for specific champions by name
export default function NameFilter({ change, setChange }: IFilter) {
    return (
        <Box
            sx={{
                width: "calc(100vw - 315px)",
                float: "right",
                textAlign: "center",
            }}
            role="NameFilter"
        >
            <ThemeProvider theme={fieldTheme}>
                <TextField
                    id="outlined-basic"
                    label="Champion name"
                    variant="outlined"
                    onChange={() => {
                        setChange(!change);
                    }}
                    sx={{ width: "50vw" }}
                    InputProps={{
                        endAdornment: (
                            <InputAdornment position="end">
                                <IconButton edge="end">
                                    <SearchIcon
                                        sx={{ color: "#CFB53B" }}
                                    ></SearchIcon>
                                </IconButton>
                            </InputAdornment>
                        ),
                    }}
                />
            </ThemeProvider>
        </Box>
    );
}
