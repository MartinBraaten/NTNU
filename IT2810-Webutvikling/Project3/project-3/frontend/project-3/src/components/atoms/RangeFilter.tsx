//Filter for filtering on champions range
import { Slider, Typography } from "@mui/material";
import Box from "@mui/material/Box";
import { useEffect } from "react";

export interface IRangeFilter {
    change: boolean | undefined;
    setChange: (change: boolean) => void;
    value: number[];
    setValue: (value: number[]) => void;
}

//inspired by https://mui.com/material-ui/react-slider/
//Component for filtering champions on attack range
export function RangeFilter({
    change,
    setChange,
    value,
    setValue,
}: IRangeFilter) {
    //Sets a state that runs a useEffect in HomePage that updates all filters when a filter is changed
    useEffect(() => {
        setChange(!change);
    }, [value]);

    //Updates values on slider and the values displayed in the text under the slider
    const HandleChange = (event: Event, newValue: number | number[]) => {
        setValue(newValue as number[]);
    };
    return (
        <Box sx={{ width: "200px" }} role="RangeFilter">
            <Typography
                align="center"
                variant="h6"
                sx={{ fontSize: "18px", fontWeight: "400" }}
            >
                Attack range:
            </Typography>
            <Slider
                role="rangeSlider"
                sx={{ color: "#cfb53b" }}
                value={value}
                onChange={HandleChange}
                valueLabelDisplay="auto"
                max={650}
                disableSwap
            />
            <Typography
                role="sliderValues"
                color="#fff"
                align="center"
                variant="h6"
                sx={{ fontSize: "18px", fontWeight: "400" }}
            >
                {value[0]} - {value[1]}
            </Typography>
        </Box>
    );
}
