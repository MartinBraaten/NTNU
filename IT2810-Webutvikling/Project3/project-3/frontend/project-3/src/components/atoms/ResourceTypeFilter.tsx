//Filter for filtering o champion tags
import {
    Autocomplete,
    TextField,
    ThemeProvider,
    Typography,
} from "@mui/material";
import Box from "@mui/material/Box";
import { useEffect, useState } from "react";
import { fieldTheme } from "../themes/themes";

//inspired by https://mui.com/material-ui/react-autocomplete/
//The options for champion resource type
const resourceTypes = [
    "Blood Well",
    "Mana",
    "Energy",
    "None",
    "Rage",
    "Courage",
    "Shield",
    "Fury",
    "Ferocity",
    "Heat",
    "Grit",
    "Crimson Rush",
    "Flow",
    "Health Cost",
];

export interface IResourceFilter {
    change: boolean | undefined;
    setChange: (change: boolean) => void;
}

//Component for filtering champions on their resource type
export default function ResourceTypeFilter({
    change,
    setChange,
}: IResourceFilter) {
    const [value, setValue] = useState<string | null>(null);
    //Sets a state that runs a useEffect in HomePage that updates all filters when a filter is changed
    useEffect(() => {
        setChange(!change);
    }, [value]);

    return (
        <Box sx={{ width: "200px" }} role="ResourceTypeFilter">
            <Typography
                align="center"
                variant="h6"
                sx={{ fontSize: "18px", fontWeight: "400" }}
            >
                Resource type:
            </Typography>
            <ThemeProvider theme={fieldTheme}>
                <Autocomplete
                    value={value}
                    disablePortal
                    id="resourceField"
                    options={resourceTypes}
                    onChange={(event: any, tag: string | null) => {
                        setValue(tag);
                    }}
                    sx={{ width: "200px", color: "#fff" }}
                    renderInput={(params) => (
                        <TextField
                            role="ChampionResource"
                            {...params}
                            label="Resource type"
                        />
                    )}
                />
            </ThemeProvider>
        </Box>
    );
}
