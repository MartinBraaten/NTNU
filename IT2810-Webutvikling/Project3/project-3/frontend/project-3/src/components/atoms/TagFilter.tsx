//Filter for filtering o champion tags
import {
    Autocomplete,
    TextField,
    ThemeProvider,
    Typography,
} from "@mui/material";
import Box from "@mui/material/Box";
import { useEffect, useState } from "react";
import { fieldTheme } from "../themes/themes";

//inspired by https://mui.com/material-ui/react-autocomplete/
//Options for champion tag
const championTags = [
    "Marksman",
    "Support",
    "Assassin",
    "Mage",
    "Fighter",
    "Tank",
];

export interface ITagFilter {
    change: boolean | undefined;
    setChange: (change: boolean) => void;
}

//Component for filtering champions on their tag
export default function TagFilter({ change, setChange }: ITagFilter) {
    const [value, setValue] = useState<string | null>(null);

    //Sets a state that runs a useEffect in HomePage that updates all filters when a filter is changed
    useEffect(() => {
        setChange(!change);
    }, [value]);
    return (
        <Box sx={{ width: "200px" }} role="TagFilter">
            <Typography
                align="center"
                variant="h6"
                sx={{ fontSize: "18px", fontWeight: "400" }}
            >
                Tag:
            </Typography>

            <ThemeProvider theme={fieldTheme}>
                <Autocomplete
                    value={value}
                    disablePortal
                    id="tagSelect"
                    options={championTags}
                    onChange={(event: any, tag: string | null) => {
                        setValue(tag);
                    }}
                    sx={{ width: "200px", color: "#fff" }}
                    renderInput={(params) => (
                        <TextField role="ChampionTag" {...params} label="Tag" />
                    )}
                />
            </ThemeProvider>
        </Box>
    );
}
