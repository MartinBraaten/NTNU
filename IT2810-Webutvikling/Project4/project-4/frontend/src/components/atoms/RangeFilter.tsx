//Filter for filtering on champions range
import { Slider, Typography } from "@mui/material";
import Box from "@mui/material/Box";
import { useDispatch, useSelector } from "react-redux";
import { selectFilters, setFilter } from "../store/FiltersReducer";

//inspired by https://mui.com/material-ui/react-slider/
//Component for filtering champions on attack range
export function RangeFilter() {
    const dispatch = useDispatch();
    const filters = useSelector(selectFilters);

    return (
        <Box sx={{ width: "200px" }} role="RangeFilter">
            <Typography
                align="center"
                variant="h6"
                sx={{ fontSize: "18px", fontWeight: "400" }}
            >
                Attack range:
            </Typography>
            <Slider
                role="rangeSlider"
                sx={{ color: "#cfb53b" }}
                value={filters.range}
                onChange={(event: any, range: number | number[]) => {
                    dispatch(setFilter({ field: "range", value: range }));
                }}
                valueLabelDisplay="auto"
                max={650}
                disableSwap
            />
            <Typography
                role="sliderValues"
                color="#fff"
                align="center"
                variant="h6"
                sx={{ fontSize: "18px", fontWeight: "400" }}
            >
                {filters.range[0]} - {filters.range[1]}
            </Typography>
        </Box>
    );
}
