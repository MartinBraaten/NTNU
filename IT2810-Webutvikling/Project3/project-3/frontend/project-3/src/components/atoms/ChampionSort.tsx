//Filter for filtering o champion tags
import {
    Autocomplete,
    TextField,
    ThemeProvider,
    Typography,
} from "@mui/material";
import Box from "@mui/material/Box";
import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { selectChampions, sortChampions } from "../store/ChampionReducer";
import { removeFilter } from "../store/FilteredChampionsReducer";
import { fieldTheme } from "../themes/themes";
import { IFilter } from "./NameFilter";

//inspired by https://mui.com/material-ui/react-autocomplete/
//The options for ordering champions
const orderTypes = ["Alphabetical", "Attack range"];

//Component that lets you choose which order to display the champions in
export default function ChampionSort({ change, setChange }: IFilter) {
    const [value, setValue] = useState<string | null>(null);
    const dispatch = useDispatch();
    const allChampions = useSelector(selectChampions);

    //Set value og textfield to the chosen alternative
    const handleChange = (event: any, order: string | null) => {
        setValue(order);
    };
    //Order the champions everytime a new order is chosen
    useEffect(() => {
        dispatch(sortChampions(value));
        dispatch(removeFilter(allChampions));
        setChange(!change);
    }, [value]);

    return (
        <Box sx={{ width: "200px" }} role="ChampionSort">
            <Typography
                align="center"
                variant="h6"
                sx={{ fontSize: "18px", fontWeight: "400" }}
            >
                Sort by:
            </Typography>
            <ThemeProvider theme={fieldTheme}>
                <Autocomplete
                    value={value}
                    disablePortal
                    options={orderTypes}
                    id="championsSort"
                    onChange={handleChange}
                    sx={{ width: "200px", color: "#fff" }}
                    renderInput={(params) => (
                        <TextField {...params} label="Sort order" />
                    )}
                />
            </ThemeProvider>
        </Box>
    );
}
