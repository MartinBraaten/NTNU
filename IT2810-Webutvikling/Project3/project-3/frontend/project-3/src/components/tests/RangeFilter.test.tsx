/* eslint-disable testing-library/await-async-query */
/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-container */
import { render } from "@testing-library/react";

// eslint-disable-next-line testing-library/no-dom-import
import { screen } from "@testing-library/dom";
import { RangeFilter } from "../atoms/RangeFilter";

test("Renders correctly", () => {
    function mockFunction(prop: boolean) { }
    function mockFunction2(prop: number | number[]) { }
    let { container } = render(<RangeFilter change={undefined} setChange={mockFunction} value={[]} setValue={mockFunction2} />);
    expect(container).toMatchSnapshot();
});

test("Display correct numbers", () => {
    function mockFunction(prop: boolean) { }
    function mockFunction2(prop: number | number[]) { }
    render(<RangeFilter change={undefined} setChange={mockFunction} value={[0, 650]} setValue={mockFunction2} />);

    let textComponent = screen.getByRole(/^sliderValues/i);
    expect(textComponent.innerHTML).toMatch("0 - 650");
});

test("Updates displayed values on slider change", () => {

    //cannot run because of hooks used in component (useState)
/*
    const component = TestRenderer.create(<RangeFilter />);

    component.root.findByType(Slider).props.onChange(null, [50, 175]);

    component.root.findByType(Slider).props.value = [50, 175];

    let textComponent = screen.getByRole(/^sliderValues/i);
    expect(textComponent.innerHTML).toMatch("50 - 175");
*/
});

