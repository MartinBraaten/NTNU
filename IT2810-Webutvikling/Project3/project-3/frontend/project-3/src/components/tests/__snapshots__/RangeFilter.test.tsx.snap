// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-1domaf0"
    role="RangeFilter"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-fff52w-MuiTypography-root"
    >
      Attack range:
    </h6>
    <span
      class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-sizeMedium css-10drvo-MuiSlider-root"
      role="rangeSlider"
    >
      <span
        class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
      />
      <span
        class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
      />
    </span>
    <h6
      class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-gyuy0t-MuiTypography-root"
      role="sliderValues"
    >
       - 
    </h6>
  </div>
</div>
`;
